src/user.rs

```diff
@@ -20,107 +19,102 @@ pub(crate) struct User {
 	/// Whether the user is enabled
 	pub(crate) enabled: bool,
 	/// The user's preferred username
-	pub(crate) preferred_username: StringOrBytes,
-	/// The user's LDAP ID
+	pub(crate) preferred_username: Option<StringOrBytes>,
+	/// The user's external (non-Zitadel) ID
 	pub(crate) external_user_id: StringOrBytes,
 }
 
 impl User {
-	/// Convert the agnostic user to a Zitadel user
-	pub fn to_zitadel_user(&self, feature_flags: &FeatureFlags, idp_id: &str) -> ZitadelUser {
-		ZitadelUser {
-			user_data: self.clone(),
-			needs_email_verification: feature_flags.is_enabled(FeatureFlag::VerifyEmail),
-			needs_phone_verification: feature_flags.is_enabled(FeatureFlag::VerifyPhone),
-			idp_id: feature_flags.contains(&FeatureFlag::SsoLogin).then(|| idp_id.to_owned()),
-		}
-	}
-}
+	/// Convert a Zitadel user to our internal representation
+	pub fn try_from_zitadel_user(user: HumanUser, external_id: String) -> Result<Self> {
+		let first_name = user
+			.profile()
+			.and_then(|profile| profile.given_name())
+			.ok_or(anyhow!("Missing first name for {}", external_id))?
+			.clone();
 
-/// Crate-internal representation of a Zitadel user
-#[derive(Clone, Debug)]
-pub struct ZitadelUser {
-	/// Details about the user
-	pub(crate) user_data: User,
-
-	/// Whether the user should be prompted to verify their email
-	pub(crate) needs_email_verification: bool,
-	/// Whether the user should be prompted to verify their phone number
-	pub(crate) needs_phone_verification: bool,
-	/// The ID of the identity provider to link with, if any
-	pub(crate) idp_id: Option<String>,
-}
+		let last_name = user
+			.profile()
+			.and_then(|profile| profile.family_name())
+			.ok_or(anyhow!("Missing last name for {}", external_id))?
+			.clone();
 
-impl ZitadelUser {
-	/// Get a display name for the user
-	pub(crate) fn get_display_name(&self) -> String {
-		format!("{}, {}", self.user_data.last_name, self.user_data.first_name)
-	}
+		let email = user
+			.email()
+			.and_then(|human_email| human_email.email())
+			.ok_or(anyhow!("Missing email address for {}", external_id))?
+			.clone();
 
-	/// Return the name to be used in logs to identify this user
-	pub(crate) fn log_name(&self) -> String {
-		format!("email={}", &self.user_data.email)
+		let phone = user.phone().and_then(|human_phone| human_phone.phone());
+
+		let external_user_id = match BASE64_STANDARD.decode(external_id.clone()) {
```

Comment:
I've had to think really hard about how this all works while trying to write a migration guide for this, and realized that the current implementation simply cannot work for sources that have non-UTF8 external user IDs (read: AD).

The issue is that base64 encoding values **apparently does *not* preserve alphabetic ordering**. This is an issue, because it means that the order our source returns will be different from the one Zitadel returns if Zitadel deals with base64-encoded values and the source does not, thereby completely breaking the sync algorithm.

It's unfortunate that I realized this this far down the line...

We will need to handle this completely differently. I'd propose already encoding binary IDs at the source level (i.e., when parsing IDs from *any* source), and only ever handling encoded values for IDs elsewhere in the codebase. This guarantees that the sort order is consistent. We should also never pass `ByteOrString` enums into the public interface of the `user` module - Zitadel should only ever deal with `String`s anyway, so this will clean up the codebase, too.

This will also mean that Zitadel will only ever see encoded values for IDs, making the decode step completely unnecessary.

In addition, since base64 would prevent us from ever using source APIs to sort incoming users instead of doing it on the application end, incurring pointless memory overhead, we should not use base64 encoding to do this, but instead an encoding scheme that *does* preserve alphabetical order. I'm not aware of any off the top of my head, but I am sure they exist.

This also gives us a clean migration path, as we can write a separate command that goes through all users of a Zitadel instance and re-encodes their IDs in that encoding, which must then be run on any Zitadel instances that have previously been synced to using this tool.

On the up side, since the ldap attribute parsing implementation [was much more conscientious than I remembered](https://github.com/famedly/famedly-sync/blob/6f6db25e0d2688beed547b6c5605e3df987ce3c6/src/sources/ldap.rs#L112), there should be no instances with mixed binary/string IDs, so such a tool can have subcommands for binary/string mode re-encoding, and we should not run any particular risk of accidentally decoding a value that was originally intended to be a string.

Unfortunately, Zitadel [still requires this broken way of passing through user IDs](https://github.com/zitadel/zitadel/blob/main/internal/idp/providers/ldap/session.go#L289) for IDP IDs, so we will need to decode the value we just encoded just to re-encode it with a different encoding scheme if the original value happens to not be valid UTF8. At least it helps me with my impostor syndrome to know that other programmers clearly stumble over the same quirks of ldap.

Sorry for the delay in all this, this has been *so* much more complex than it seems at face value.

Reply:
Thank you for the extensive problem description. Let's talk about it in our sync tomorrow.

Regarding the encoding scheme, I think base32hex is a good option here: https://www.rfc-editor.org/rfc/rfc4648#page-10